### ISSUE TITLE
JSONata 2.0.0 - Is variable binding no longer working within the object creator?

### ISSUE THREAD

**Issue by oatybiscuit** [jsonata-js/jsonata#607]
Using an expression with {} at the end, containing variable bindings to capture computed key-values so as to generate another value from them. Works perfectly in V1.8.6 but noticed today it does not work in V2.0.0.

Is this a feature, a breaking change, something that was never supposed to work (but did), or just my error as I should not be doing this?

A simplified example-

{"one": $a:=1,
 "two": $b:=2,
 "sum": $a+$b,
 "true": $tt:=true,
 "false": $tf:=false,
 "logic": $tt and $not($tf)
}

In V1.8.6 I get
{
  "one": 1,
  "two": 2,
  "sum": 3,
  "true": true,
  "false": false,
  "logic": true
}

In V2.0.0 I get
{
  "one": 1,
  "two": 2,
  "true": true,
  "false": false,
  "logic": false
}

**Comment by markmelville** [comment]
I don't think using variable binding within object creation is ever demonstrated in the docs. All uses of the `:=` operation are done in a statement `( ; ; )`. So I'd say that what you are doing here would be be something with an undefined behavior.

"The scope of a variable is limited to the 'block' in which it was bound." So it may be that in v1 the block was the object, but in v2 the block is the value of each property. Thus $a and $b are already out of scope when evaluating "$a+$b".

This slight variation work in both versions:
```
(
    $a:=1;
    $b:=2;
    $tt:=true;
    $tf:=false;
    {
        "one": $a,
        "two": $b,
        "sum": $a+$b,
        "true": $tt,
        "false": $tf,
        "logic": $tt and $not($tf)
    }
)
```

**Comment by oatybiscuit** [comment]
Thank you for your feedback, much appreciated. I have 'fixed' (made V2 proof) my code as per your comments and that certainly now works for both versions.

My further thoughts:-

As documented:
Variable declaration occurs in () blocks using := assignment, and the scope is the enclosing block.
Variables can also be declared and 'assigned' using the path positional and context markers # and @. I use #$i regularly. Scope here lasts only for the statement line.

Syntax with undefined behaviour:
I have experimented a little more - this is using the 'invoice' test data, so very easily replicated

`Account.Order.Product.{$k:="Total @ " & $string($q:=Quantity) : $q*Price, "how many": $q, "key is": $k}`

So, two variables $k and $q. Declared, assigned, and then used, and both apparently work, and for both versions.

$q is 'declared' in a "block", strictly the function call for $string, but it becomes available both for that key value and for the next key value. If the variable scope is the enclosing block only, then I assume that the () are part of the function call and not a block.
$k is declared in the {} but is here available for the last key-value. If this works, does that mean that the {} object creator is the block?

Since Quantity is a number, I use $string() to convert to a string for the key, however the & operator will force a number to a string, so the key value construct could be "Total @ " & Quantity. Whilst $q is not available, this still allows $k to execute.

Using "Total @ " & ($q:=Quantity) also works for $k, as $q is declared and assigned, and quantity returned from the () then cast to a string, but the scope of $q I assume is now limited to the small block and thus does not appear elsewhere.
If, however, I add back the redundant "$string" making the block now a function call, $q becomes available to all the key expressions.

I assume that within ( ; ; ) code blocks, compilation and execution formally accommodates and executes variable declaration, binding and scope, however elsewhere in the expression path any variable declaration (other than for # and @) occurs by accident due to (but not excluded to) function calls and the like. Since V2 has made a breaking change to the way that compilation and evaluation is executed, undefined behaviour has the potential to change!

Oh, and if anyone has sight of a BNF document or similar for JSONata syntax, I would be most grateful for a copy!

**Comment by andrew-coleman** [comment]
Thanks for raising this, and for the discussion.  This would definitely fall into the category of "undefined behaviour".  There is no guaranteed order of execution for each of the name/value pairs of an object.  Indeed, a JSONata processor could choose to evaluate them in parallel (this Javascript implementation doesn't, but the Go/Java ones might).  I know that the [Go implementation](https://github.com/blues/jsonata-go) will iterate through the object properties in an undefined order, and we had to be careful when writing the tests not to assume any particular order (e.g. with wildcards or `$each`).  The only structure that imposes order of execution is the block `( ... ; ...  ; ... )`, as observed above.

**Comment by oatybiscuit** [comment]
Thank you - undefined behaviour, with a personal note not to assume any particular order of execution except for the block construct, which is executed line by line.
